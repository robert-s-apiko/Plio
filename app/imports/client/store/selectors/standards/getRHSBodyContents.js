import { createSelector } from 'reselect';

import { getStandardBookSectionsByIds } from '../standardBookSections';
import { getStandardTypesByIds } from '../standardTypes';
import { getUsersByIds } from '../users';
import pickDocuments from '../../../../api/helpers/pickDocuments';

/*
  interface improvementPlan {
    owner: ID
  }

  selector({
    sectionId: ID,
    typeId: ID,
    owner: ID,
    notify: [...ID],
    improvementPlan
  }) => (
    standardBookSectionsByIds: Object,
    standardTypesByIds: Object,
    usersByIds: Object
  ) => Object
*/

const selector = ({
  sectionId,
  typeId,
  owner,
  notify,
  improvementPlan = {},
} = {}) => (
  standardBookSectionsByIds,
  standardTypesByIds,
  usersByIds,
) => ({
  section: standardBookSectionsByIds[sectionId],
  type: standardTypesByIds[typeId],
  owner: usersByIds[owner],
  improvementPlan: {
    ...improvementPlan,
    owner: usersByIds[improvementPlan.owner],
  },
  notify: pickDocuments(['_id', 'profile', 'emails'], usersByIds, notify),
});

export default props => createSelector([
  getStandardBookSectionsByIds,
  getStandardTypesByIds,
  getUsersByIds,
], selector(props));
