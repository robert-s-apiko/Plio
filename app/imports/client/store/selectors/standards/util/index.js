export { default as withUncategorized } from './withUncategorized';
