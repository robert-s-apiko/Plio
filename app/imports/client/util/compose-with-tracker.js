import { Tracker } from 'meteor/tracker';
import { compose } from 'ramda';

function getTrackerLoader(loaderFunc) {
  return (props, onData, env) => {
    let trackerCleanup = () => null;

    const handler = Tracker.nonreactive(() =>
      Tracker.autorun(() => {
        // Store clean-up function if provided.
        trackerCleanup = loaderFunc(props, onData, env) || (() => null);
      }));

    return () => {
      trackerCleanup();
      return handler.stop();
    };
  };
}

function composeWithTracker(loadFunc) {
  return function (component) {
    return compose(getTrackerLoader(loadFunc))(component);
  };
}

export default composeWithTracker;
