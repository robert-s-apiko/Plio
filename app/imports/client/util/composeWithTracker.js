import { Tracker } from 'meteor/tracker';
import { compose } from '@storybook/react-komposer';

function getTrackerLoader(loaderFunc) {
  return (props, onData, env) => {
    let trackerCleanup = () => null;

    const handler = Tracker.nonreactive(() =>
      Tracker.autorun(() => {
        // Store clean-up function if provided.
        trackerCleanup = loaderFunc(props, onData, env) || (() => null);
      }));

    return () => {
      trackerCleanup();
      return handler.stop();
    };
  };
}

function composeWithTracker(loadFunc, options) {
  return function (component) {
    // explicitly pass { withRef: false }
    // otherwise react will throw error
    // that the component tries to access
    // a ref of the stateless component
    const opts = Object.assign({ withRef: false }, options);

    return compose(getTrackerLoader(loadFunc), opts)(component);
  };
}

export default composeWithTracker;
