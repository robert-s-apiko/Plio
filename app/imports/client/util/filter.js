import { ifElse, equals, compose, filter, nthArg, complement } from 'ramda';

import { isDeleted } from './object-relation';

// (filterDeleted: Boolean, items: Array) => Array
export const filterByDeleted = ifElse(
  equals(true),
  compose(filter(isDeleted), nthArg(1)),
  compose(filter(complement(isDeleted)), nthArg(1)),
);
