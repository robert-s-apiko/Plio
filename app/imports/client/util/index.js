export { default as composeWithTracker } from './composeWithTracker';
