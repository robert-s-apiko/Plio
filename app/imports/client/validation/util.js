import { FORM_ERROR } from 'final-form';
import { forEachObjIndexed, curry, isNil } from 'ramda';

export const required = curry((label, value) => !isNil(value) ? undefined : `${label} required`);

export const validate = rules => (values) => {
  const errors = new Set();

  forEachObjIndexed((value, key) => {
    if (rules[key]) {
      const error = rules[key](value, values);
      if (error) errors.add(error);
    }
  }, values);

  if (!errors.size) return undefined;

  return { [FORM_ERROR]: [...errors].join('\n') };
};
