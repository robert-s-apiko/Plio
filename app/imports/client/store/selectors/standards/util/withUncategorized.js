import { set, filter, propEq, append, compose, map } from 'ramda';
import { lenses, getStandardsLength } from '../../../../util';

// (prop: String, standards: Array) => (item: Object) => Object
export const setFilteredStandards = (prop, standards) => item => set(
  lenses.standards,
  filter(propEq(prop, item._id), standards),
  item,
);

// (prop: String, uncategorized: Object, standards: Array, items: Array) => Array
export default (prop, uncategorized, standards, items) => compose(
  filter(getStandardsLength),
  append(uncategorized),
  map(setFilteredStandards(prop, standards)),
)(items);
