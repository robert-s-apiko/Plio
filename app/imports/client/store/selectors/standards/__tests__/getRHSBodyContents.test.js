import getRHSBodyContents from '../getRHSBodyContents';
import { getInitialState } from '../../../util/tests';

describe('getRHSBodyContents', () => {
  it('matches snapshot', () => {
    const state = getInitialState({});
    const props = {
      sectionId: 8,
      typeId: 9,
      owner: 6,
      notify: [6],
      improvementPlan: {
        owner: 6,
      },
    };

    expect(getRHSBodyContents(props)(state)).toMatchSnapshot();
  });
});
