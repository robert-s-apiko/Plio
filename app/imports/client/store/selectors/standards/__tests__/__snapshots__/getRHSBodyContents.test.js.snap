// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRHSBodyContents matches snapshot 1`] = `
Object {
  "improvementPlan": Object {
    "owner": Object {
      "_id": 6,
      "roles": Object {
        "7": Array [
          "create-update-delete-standards",
          "view-team-actions",
          "invite-users",
          "delete-users",
          "edit-user-roles",
          "change-org-settings",
        ],
      },
    },
  },
  "notify": Array [
    Object {
      "_id": 6,
      "emails": undefined,
      "profile": undefined,
    },
  ],
  "owner": Object {
    "_id": 6,
    "roles": Object {
      "7": Array [
        "create-update-delete-standards",
        "view-team-actions",
        "invite-users",
        "delete-users",
        "edit-user-roles",
        "change-org-settings",
      ],
    },
  },
  "section": Object {
    "_id": 8,
  },
  "type": Object {
    "_id": 9,
  },
}
`;
